### PR TITLE
keystore and account support for multi key

### PR DESCRIPTION
**Pull request by shayzluf**
[Part of] #2025

---

# Update validator accounts creation and keystore to support multiple keys

following the implementation guidelines https://github.com/prysmaticlabs/prysm/issues/2025

### CODE REVIEW

File: shared/keystore/keystore.go [hunk 1]
```diff
@@ -80,6 +81,38 @@ func (ks Store) GetKey(filename, password string) (*Key, error) {
...
+		if f.Mode().IsRegular() && strings.Contains(n, strings.TrimPrefix(fileprefix, "/")) {
+			keyjson, err := ioutil.ReadFile(filePath)
+			if err != nil {
+				return nil, err
+			}
+			key, err := DecryptKey(keyjson, password)
+			if err != nil {
+				return nil, err
+			}
+			keys = append(keys, key)
```
Comment (skillful-alex) [inline shared/keystore/keystore.go]: Can the keys be stored not in the array, but in the map, using the public key as the map key?
Or at least check the key for uniqueness.

Now, if there is the same key in different files (copying, backups, etc.), then a double vote is likely ...

Comment (shayzluf) [reply shared/keystore/keystore.go]: good observation.
Thanks

Comment (prestonvanloon) [reply shared/keystore/keystore.go]: Are you planning to address this comment? @shayzluf

**Comment by prestonvanloon** [comment]
Overall LGTM.

I'll verify the following and update this comment:
- [ ] Runs locally without issue
- [ ] Runs in cluster without issue

File: shared/keystore/keystore.go (hunk 1 above)
Comment (terenc3t) [reply shared/keystore/keystore.go]: do we want to address this?

File: validator/accounts/account.go [hunk 2]
```diff
@@ -39,13 +40,10 @@ func VerifyAccountNotExists(directory string, password string) error {
...
-	// First, if the keystore already exists, throws an error as there can only be
-	// one keystore per validator client.
-	if err := VerifyAccountNotExists(directory, password); err != nil {
-		return fmt.Errorf("validator account exists: %v", err)
-	}
-	shardWithdrawalKeyFile := directory + params.BeaconConfig().WithdrawalPrivkeyFileName
-	validatorKeyFile := directory + params.BeaconConfig().ValidatorPrivkeyFileName
+	t := time.Now()
+
+	shardWithdrawalKeyFile := directory + params.BeaconConfig().WithdrawalPrivkeyFileName + t.Format("20060102150405")
```
Comment (terenc3t) [inline validator/accounts/account.go]: as @nisdas pointed out, is there a better way to mark them? maybe use the pubkey for suffix ? 
Having the time doesnt really help in differentiating them...

Comment (prestonvanloon) [reply validator/accounts/account.go]: Pubkey substring sounds good to me. Maybe take the first 12 characters of the key. 

The full key is way too long

File: shared/keystore/keystore.go (hunk 1 above)
Comment (shayzluf) [reply shared/keystore/keystore.go]: done

Comment (prestonvanloon) [reply shared/keystore/keystore.go]: How is this done? It is still possible that the `keys` array could have duplicate keys

File: shared/keystore/keystore.go [hunk 3]
```diff
@@ -80,6 +81,38 @@ func (ks Store) GetKey(filename, password string) (*Key, error) {
...
+		return nil, err
+	}
+	keys := []*Key{}
+	i := 0
+	for _, f := range files {
+		n := f.Name()
+		filePath := filepath.Join(directory, n)
+		filePath = filepath.Clean(filePath)
+		if f.Mode().IsRegular() && strings.Contains(n, strings.TrimPrefix(fileprefix, "/")) {
+			keyjson, err := ioutil.ReadFile(filePath)
```
Comment (prestonvanloon) [inline shared/keystore/keystore.go]:
```suggestion
                        // #nosec G304
			keyjson, err := ioutil.ReadFile(filePath)
```

Reading a file by variable name is a low security risk, so you need to add the exclusion comment.

File: shared/keystore/keystore.go (hunk 1 above)
Comment (shayzluf) [reply shared/keystore/keystore.go]: I merged both pr into #2069 and addressed it there. sorry that I didn't notice that the comment is on an older pr

File: validator/accounts/account.go (hunk 2 above)
Comment (shayzluf) [reply validator/accounts/account.go]: I merged this pr into #2069 and resolved it there.

**Comment by prestonvanloon** [comment]
Running in cluster now, passes the smoke test. Merging!

**Pull request merged**